### PR TITLE
Add host entitlements AND configurable AppArmor profiles

### PR DESCRIPTION
**Pull request by n4ss**
This PR adds `host.*` entitlements, regrouping `host.devices.*` and `host.processes.*` entitlements.

I imported `templates` pkg from Moby, and pieces from `apparmor` pkg from Moby also but modified it. AppArmor profile generation still needs love as I only added what I needed in the template file.

### CODE REVIEW

**Comment by n4ss** [comment]
Due to a big number of failures when testing, I mark this PR as WIP.

Following things need to be fixed:
- apparmor profile template's syntax is being fixed as well as `ProfileData` struct
- need to initialize the apparmor.ProfileData struct with the right default values

**Comment by n4ss** [comment]
PR is ready for review @riyazdf if you like :)

Will now focus on integration tests as a more images now work with a combination of entitlements.

File: apparmor/apparmor.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	"io"
+	"os"
+	"path"
+
+	"github.com/docker/libentitlement/templates"
+)
+
+var (
+	// profileDirectory is the file store for apparmor profiles and macros.
+	profileDirectory = "/etc/apparmor.d"
```
Comment (riyazdf) [inline apparmor/apparmor.go]: is this the same for most platforms?  If we aren't sure, we should make this a `defaultProfileDirectory` and have a function that allows users to override this

File: apparmor/apparmor.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+	// Capabilities defines the capabilities setup we want, see CapabiltitiesSetup type definition
+	Capabilities CapabilitiesSetup
+
+	// Files defines the files access setup we want, see FilesSetup type definition
+	Files FilesSetup
+}
+
+// NewProfileData creates an ProfileData object with its name.
+func NewProfileData(name string) *ProfileData {
```
Comment (riyazdf) [inline apparmor/apparmor.go]: nit: can we rename this to `NewEmptyProfileData`?  I'm wondering if we want to keep more of the fields like `Capabilities` unexported and have getters/setters, but we can decide later.

File: defaults/host-ents.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+	for _, allowedMount := range allowedMounts {
+		if reflect.DeepEqual(mount, allowedMount) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func hostDevicesNoneEntitlementEnforce(profile secprofile.Profile) (secprofile.Profile, error) {
```
Comment (riyazdf) [inline defaults/host-ents.go]: could you add a quick docstring to explain this entitlement and the other new ones?

File: defaults/host-ents.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+	hostDevicesNoneEntitlement  = entitlement.NewVoidEntitlement(HostDevicesNoneEntFullID, hostDevicesNoneEntitlementEnforce)
+	hostDevicesViewEntitlement  = entitlement.NewVoidEntitlement(HostDevicesViewEntFullID, hostDevicesViewEntitlementEnforce)
+	hostDevicesAdminEntitlement = entitlement.NewVoidEntitlement(HostDevicesAdminEntFullID, hostDevicesAdminEntitlementEnforce)
+
+	hostProcessesNoneEntitlement  = entitlement.NewVoidEntitlement(HostProcessesNoneEntFullID, hostProcessesNoneEntitlementEnforce)
+	hostProcessesAdminEntitlement = entitlement.NewVoidEntitlement(HostProcessesAdminEntFullID, hostProcessesAdminEntitlementEnforce)
+)
+
+var (
+	allowedMounts = []specs.Mount{
```
Comment (riyazdf) [inline defaults/host-ents.go]: can this be a `const`?  Also, can we rename this to something more descriptive to represent host devices and sysfs?

File: defaults/host-ents.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+	}
+
+	return ociProfile, nil
+}
+
+func removeReadOnlyFlagMount(mount specs.Mount) specs.Mount {
+	readWriteMount := mount
+	for index, option := range readWriteMount.Options {
+		if option == "ro" {
+			readWriteMount.Options = append(readWriteMount.Options[:index], readWriteMount.Options[index+1:]...)
```
Comment (riyazdf) [inline defaults/host-ents.go]: since this is used for admin, do we want to remove the mount or actually allow it to be read-write?

File: defaults/network-ents.go [hunk 6]
```diff
@@ -97,24 +97,24 @@ func networkUserEntitlementEnforce(profile secprofile.Profile) (secprofile.Profi
...
+	//syscallsWithArgsToAllow := map[types.Syscall][]specs.LinuxSeccompArg{
+	//	SysSetsockopt: {
+	//		{
+	//			Index: 2,
+	//			Value: syscall.SO_DEBUG,
+	//			Op:    specs.OpNotEqual,
+	//		},
+	//	},
+	//}
+	//ociProfile.AllowSyscallsWithArgs(syscallsWithArgsToAllow)
```
Comment (riyazdf) [inline defaults/network-ents.go]: why is this commented out?  Is this no longer needed?

File: defaults/network-ents.go [hunk 7]
```diff
@@ -134,7 +134,7 @@ func networkProxyEntitlementEnforce(profile secprofile.Profile) (secprofile.Prof
 	capsToRemove := []types.Capability{CapNetAdmin}
 	ociProfile.RemoveCaps(capsToRemove...)
 
-	capsToAdd := []types.Capability{CapNetBroadcast, CapNetRaw, CapNetBindService}
```
Comment (riyazdf) [inline defaults/network-ents.go]: I'm curious - what's the reasoning behind this change?

File: defaults/security-ents.go [hunk 8]
```diff
@@ -37,13 +37,17 @@ func securityConfinedEntitlementEnforce(profile secprofile.Profile) (secprofile.
 	}
 
 	capsToRemove := []types.Capability{
-		CapMacAdmin, CapMacOverride, CapDacOverride, CapDacReadSearch, CapSetfcap, CapSetfcap, CapSetuid, CapSetgid,
+		CapMacAdmin, CapMacOverride, CapDacOverride, CapDacReadSearch, CapSetpcap, CapSetfcap, CapSetuid, CapSetgid,
```
Comment (riyazdf) [inline defaults/security-ents.go]: same here: what's the reason for this change?  Should we be removing both?

File: defaults/security-ents.go [hunk 9]
```diff
@@ -37,13 +37,17 @@ func securityConfinedEntitlementEnforce(profile secprofile.Profile) (secprofile.
...
 		CapSysPtrace, CapFsetid, CapSysModule, CapSyslog, CapSysRawio, CapSysAdmin, CapLinuxImmutable,
 	}
 	ociProfile.RemoveCaps(capsToRemove...)
 
 	syscallsToBlock := []types.Syscall{
-		SysPtrace, SysArchPrctl, SysPersonality, SysPersonality, SysSetuid, SysSetgid, SysPrctl, SysMadvise,
+		SysPtrace, SysArchPrctl, SysPersonality,
+		// SysSetuid,
+		// SysSetgid,
+		// SysPrctl,
```
Comment (riyazdf) [inline defaults/security-ents.go]: should we remove these?

File: secprofile/oci-profile.go [hunk 10]
```diff
@@ -44,8 +45,6 @@ func (p *OCIProfile) AddCaps(capsToAdd ...types.Capability) {
 		p.OCI.Process.Capabilities.Inheritable = addCapToList(p.OCI.Process.Capabilities.Inheritable, capStr)
 		p.OCI.Process.Capabilities.Permitted = addCapToList(p.OCI.Process.Capabilities.Permitted, capStr)
 
-		// Should be updated automatically if the previous masks are set
-		p.OCI.Process.Capabilities.Ambient = addCapToList(p.OCI.Process.Capabilities.Ambient, capStr)
```
Comment (riyazdf) [inline secprofile/oci-profile.go]: do we still need this?

File: apparmor/apparmor.go (hunk 1 above)
Comment (n4ss) [inline apparmor/apparmor.go]: Agree with this, will fix

File: apparmor/apparmor.go (hunk 2 above)
Comment (n4ss) [inline apparmor/apparmor.go]: As discussed, exposing them might prove useful but TBD

Comment (n4ss) [inline apparmor/apparmor.go]: Will rename

File: defaults/host-ents.go (hunk 3 above)
Comment (n4ss) [inline defaults/host-ents.go]: Will fix, good catch and will update documentation too

File: defaults/host-ents.go (hunk 4 above)
Comment (n4ss) [inline defaults/host-ents.go]: slices can't be const sadly :(

File: defaults/host-ents.go (hunk 5 above)
Comment (n4ss) [inline defaults/host-ents.go]: As discussed, mounts are RW by default and have the "ro" option if meant to be mounted as read-only so admins can now have sysfs etc.. mounted as RW. Will add docstring comment for this.

File: defaults/network-ents.go (hunk 6 above)
Comment (n4ss) [inline defaults/network-ents.go]: Forgot to remove it after my tests, will remove sorry.

File: defaults/security-ents.go [hunk 11]
```diff
@@ -112,15 +123,19 @@ func securityAdminEntitlementEnforce(profile secprofile.Profile) (secprofile.Pro
...
 	ociProfile.AddCaps(capsToAdd...)
 
 	syscallsToAllow := []types.Syscall{
-		SysPtrace, SysArchPrctl, SysPersonality, SysSetuid, SysSetgid, SysPrctl, SysMadvise,
+		SysPtrace, SysArchPrctl, SysPersonality, SysSetuid, SysSetgid, SysPrctl, SysMadvise, SysMount, SysInitModule,
+		SysFinitModule, SysSetns, SysClone, SysUnshare,
 	}
 	ociProfile.AllowSyscalls(syscallsToAllow...)
 
+	ociProfile.OCI.Linux.ReadonlyPaths = []string{}
```
Comment (riyazdf) [inline defaults/security-ents.go]: do we need this?  Or does it default to empty?

File: defaults/security-ents.go (hunk 8 above)
Comment (n4ss) [inline defaults/security-ents.go]: This was a typo, we were removing `CapSetfcap` twice :)

File: defaults/security-ents.go (hunk 9 above)
Comment (n4ss) [inline defaults/security-ents.go]: Yes, also forgot these after testing.

File: secprofile/oci-profile.go (hunk 10 above)
Comment (n4ss) [inline secprofile/oci-profile.go]: Removed this as this is handled by runc afaik

File: defaults/network-ents.go (hunk 7 above)
Comment (n4ss) [inline defaults/network-ents.go]: This is a typo, a proxy needs the ability to broadcast and listen to multicast. I didn't get the opportunity to test this specific entitlement properly.

Note: `network.proxy` will be a complicated one anyway.

File: defaults/network-ents.go (hunk 6 above)
Comment (n4ss) [inline defaults/network-ents.go]: Forgot to remove it after my tests, will fix

File: defaults/security-ents.go (hunk 8 above)
Comment (n4ss) [inline defaults/security-ents.go]: it was a typo, `CapSetfcap` had a duplicate in the list.

File: defaults/security-ents.go (hunk 9 above)
Comment (n4ss) [inline defaults/security-ents.go]: Yes, will fix.

File: secprofile/oci-profile.go (hunk 10 above)
Comment (n4ss) [inline secprofile/oci-profile.go]: No, this is actually handled automatically by runc afaik.

File: defaults/security-ents.go (hunk 11 above)
Comment (n4ss) [inline defaults/security-ents.go]: In Moby this is not defaulting to empty, this guarantees that there will be no read only paths.

File: defaults/security-ents.go (hunk 8 above)
Comment (riyazdf) [inline defaults/security-ents.go]: aha, yes 👍

File: defaults/security-ents.go (hunk 11 above)
Comment (n4ss) [inline defaults/security-ents.go]: Note: `--privileged` does this among other things

File: secprofile/oci-profile.go [hunk 12]
```diff
@@ -99,34 +95,56 @@ func (p *OCIProfile) AddNamespaces(nsTypes ...specs.LinuxNamespaceType) {
...
 
-		for _, syscallRule := range p.OCI.Linux.Seccomp.Syscalls {
+	syscallNameToAllowStr := string(syscallName)
 
-			if syscallRule.Action == specs.ActAllow {
-				for _, syscallName := range syscallRule.Names {
-					if syscallName == syscallNameToAllowStr &&
-						((len(syscallArgsToAllow) == 0 && len(syscallRule.Args) == 0) ||
-							reflect.DeepEqual(syscallRule.Args, syscallArgsToAllow)) {
-						return
```
Comment (n4ss) [inline secprofile/oci-profile.go]: This was a big bug, so here is the reason behind this change:
- when allowing a set of syscalls and args (a map so unordered set), we iterate over the set and try to create a seccomp rule allowing a syscall if necessary.
- if the whitelisting rule was already there for this syscall with those args we were leaving the function instead of doing the next iteration.. so we were randomly getting incomplete seccomp configuration because of iteration order on maps.

I broke down this function into two functions, one function `AllowSyscallsWithArgs ` iterating over the syscalls set to allow and a helper `allowSyscallWithArgs` function which whitelists 1 syscall with associated args if necessary.

File: defaults/host-ents.go [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+	capsToAdd := []types.Capability{
+		CapSysAdmin,
+	}
+	ociProfile.AddCaps(capsToAdd...)
+
+	return ociProfile, nil
+}
+
+/* Implements "host.processes.none" entitlement
+ * - Activates User Namespace
```
Comment (tych0) [inline defaults/host-ents.go]: Why is `.host.processes.none` related to the user namespace? Shouldn't this be pid namespace?

File: defaults/security-ents.go [hunk 14]
```diff
@@ -68,14 +74,23 @@ func securityConfinedEntitlementEnforce(profile secprofile.Profile) (secprofile.
 	return ociProfile, nil
 }
 
+/* Implements "security.view" entitlement:
+ * - Blocked caps: CAP_SYS_ADMIN, CAP_SYS_PTRACE, CAP_SETUID, CAP_SETGID, CAP_SETPCAP, CAP_SETFCAP, CAP_MAC_ADMIN,
+ *					CAP_MAC_OVERRIDE, CAP_DAC_OVERRIDE, CAP_FSETID, CAP_SYS_MODULE, CAP_SYSLOG, CAP_SYS_RAWIO,
+ *					CAP_LINUX_IMMUTABLE
```
Comment (tych0) [inline defaults/security-ents.go]: Maybe also CAP_AUDIT_READ?

File: defaults/security-ents.go [hunk 15]
```diff
@@ -104,6 +121,15 @@ func securityViewEntitlementEnforce(profile secprofile.Profile) (secprofile.Prof
 	return ociProfile, nil
 }
 
+/* Implements "security.admin" entitlement:
+ * - Authorized caps: CAP_MAC_ADMIN, CAP_MAC_OVERRIDE, CAP_DAC_OVERRIDE, CAP_DAC_READ_SEARCH, CAP_SETPCAP, CAP_SETFCAP,
+ * 						CAP_SETUID, CAP_SETGID, CAP_SYS_PTRACE, CAP_FSETID, CAP_SYS_MODULE, CAP_SYSLOG, CAP_SYS_RAWIO,
+ *						CAP_SYS_ADMIN, CAP_LINUX_IMMUTABLE, CAP_SYS_BOOT, CAP_SYS_NICE, CAP_SYS_PACCT,
+ *						CAP_SYS_TTY_CONFIG, CAP_SYS_TIME, CAP_WAKE_ALARM
```
Comment (tych0) [inline defaults/security-ents.go]: And CAP_AUDIT_WRITE?

File: apparmor/template.go [hunk 16]
```diff
@@ -0,0 +1,72 @@
+package apparmor
+
+/* See profile language http://wiki.apparmor.net/index.php/QuickProfileLanguage
+ * This profile is the base template of Moby, customized to allow entitlements to configure
+ * network, capabilities and file accesses
+ */
+const baseCustomTemplate = `
```
Comment (n4ss) [inline apparmor/template.go]: Sure that makes sense! I'll file an issue for this and add that in another PR

File: defaults/host-ents.go (hunk 13 above)
Comment (n4ss) [inline defaults/host-ents.go]: Totally, good catch... I wanted to have `host.users.none` for the User ns and `host.processes.none` for the PID ns.

File: defaults/security-ents.go (hunk 14 above)
Comment (n4ss) [inline defaults/security-ents.go]: Indeed, will fix!

File: defaults/security-ents.go (hunk 15 above)
Comment (n4ss) [inline defaults/security-ents.go]: will fix.

**Comment by n4ss** [comment]
@tych0 LGTY?

**Comment by tych0** [comment]
Yep, although I don't know how to use github's review feature to say that.

**Comment by tych0** [comment]
LGTM (testing :)

**Comment by tych0** [comment]
found it :)
